[SPMD] Change assertions to raising value errors in distributed.py

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55254 Clang-format distributed.py
* #55253 [SPMD] Restrict DDP communication hooks to SPSD mode
* #55212 [SPMD] Error out SPMD in C++ Reducer
* #54826 [SPMD] Remove test_grad_layout_1devicemodule_2replicaperprocess
* **#54825 [SPMD] Change assertions to raising value errors in distributed.py**
* #54454 [SPMD] Error out SPMD mode

These assertions are tested in test_c10d.py

Context: https://github.com/pytorch/pytorch/pull/54454#discussion_r602657818

Differential Revision: [D27381649](https://our.internmc.facebook.com/intern/diff/D27381649/)

Commit:
[SPMD] Change assertions to raising value errors in distributed.py

These assertions are tested in test_c10d.py

Context: https://github.com/pytorch/pytorch/pull/54454#discussion_r602657818

Differential Revision: [D27381649](https://our.internmc.facebook.com/intern/diff/D27381649/)

[ghstack-poisoned]

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -395,18 +395,30 @@ def __init__(
+                raise ValueError(
+                    "DistributedDataParallel device_ids and output_device arguments "
+                    "only work with single-device GPU modules, but got "
+                    "device_ids {}, output_device {}, and module parameters {}.".format(
```
Unrelated to this PR, but should it say "...and multiple devices/CPU device for module parameters {p.device for p in module.parameters()}" ?

SciPioneer replied:
Yep, updated! Thanks!

rohan-varma reviewed (approved):
Thanks! I think ValueError is better than assertions in a library.

Commit:
Update on "[SPMD] Change assertions to raising value errors in distributed.py"


These assertions are tested in test_c10d.py

Context: https://github.com/pytorch/pytorch/pull/54454#discussion_r602657818

Differential Revision: [D27381649](https://our.internmc.facebook.com/intern/diff/D27381649/)

[ghstack-poisoned]

Commit:
Update on "[SPMD] Change assertions to raising value errors in distributed.py"


These assertions are tested in test_c10d.py

Context: https://github.com/pytorch/pytorch/pull/54454#discussion_r602657818

Differential Revision: [D27381649](https://our.internmc.facebook.com/intern/diff/D27381649/)

[ghstack-poisoned]

Commit:
Update on "[SPMD] Change assertions to raising value errors in distributed.py"


These assertions are tested in test_c10d.py

Context: https://github.com/pytorch/pytorch/pull/54454#discussion_r602657818

Differential Revision: [D27381649](https://our.internmc.facebook.com/intern/diff/D27381649/)

[ghstack-poisoned]

Commit:
Update on "[SPMD] Change assertions to raising value errors in distributed.py"


These assertions are tested in test_c10d.py

Context: https://github.com/pytorch/pytorch/pull/54454#discussion_r602657818

Differential Revision: [D27381649](https://our.internmc.facebook.com/intern/diff/D27381649/)

[ghstack-poisoned]

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -396,10 +396,13 @@ def __init__(
-            "DistributedDataParallel's input module must be on "
-            "the same type of devices, but input module parameters locate in {}."
-        ).format(distinct_device_types)
+        if len(distinct_device_types) != 1:
```
There is an assert on line 389 as well and another one in `_passing_sync_batchnorm_handle`.
